### PR TITLE
http-fetcher with basic tests

### PR DESCRIPTION
**Pull request by evanshauser**
This include the most basic fetcher, http-fetcher with a couple of tests. There are a couple constructs that illustrate possible additions to the specification and a bunch of notes for additions that need to be made to make this http-fetcher a canonical example.

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+export interface FetchResult {
+  data: any;
+  errors?: any;
+  extensions?: any;
+  context?: object;
+}
+
+export type UnsubscribeHandler = () => void;
```
Comment (helfer) [inline src/types.ts]: I think this is now different in the TC39 proposal. Do we want to try to align here?

File: tests/index.ts [hunk 2]
```diff
@@ -1,9 +1,115 @@
...
+    const fetcher = new HttpFetcher('', fetch);
+    const observable = fetcher.request({
+      query: sampleQuery,
+      operationName: 'SampleQuery',
+    });
+    //observableToPromise return a promise
+    observable.subscribe({
+      next,
+      error: (error) => assert(false),
+      complete: () => { assert(next.calledOnce); done(); },
```
Comment (helfer) [inline tests/index.ts]: Have you tried if this test can fail at all?

File: tests/index.ts [hunk 3]
```diff
@@ -1,9 +1,115 @@
...
+describe('HttpFetcher', () => {
+  it('does not need any constructor arguments', () => {
+    assert.doesNotThrow(() => new HttpFetcher());
+  });
+
+  it('calls next and then complete', (done) => {
+    const next = sinon.spy();
+    const data = {hello: 'world'};
+    fetchMock.get('*', data);
+    const fetcher = new HttpFetcher('', fetch);
```
Comment (helfer) [inline tests/index.ts]: We should make the `fetch` argument optional. We'll add a separate test for adding the argument later, which can use `sinon.spy` to check that the function is getting called.

File: src/link.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        observable.subscribe({
+          next: resolve,
+          error: reject,
+        });
+      });
+    };
+  }
+
+  private static buildLinkChain(links: ApolloLink[]): NextLink {
+    const _links = [...links];
```
Comment (evanshauser) [inline src/link.ts]: This way of creating the connections between the links makes a new array for each request. It mirrors the middleware and afterware code from the netowrk interface of `apollo-client`(also in `apollo-fetch`) the `next` function on demand. We could create a closure and keep an index that we increment

File: src/setContextLink.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  ApolloLink,
+} from './link';
+
+export default class SetContextLink extends ApolloLink {
+
+  constructor (private context?: any) {
+    super();
+  }
+
+  public request(operation: Operation, forward: NextLink): Observable<FetchResult> {
```
Comment (jbaxleyiii) [inline src/setContextLink.ts]: @evanshauser I'm concerned about the current `forward` pattern since it could be optional? I'm a little confused at forward and how we even do things like enforcing passing a forward into a forward?

File: src/link.ts [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+export function execute(link: ApolloLink, operation: GraphQLRequest): Observable<FetchResult> {
+  validateOperation(operation);
+
+  if (operation.context === undefined) {
+    operation.context = {};
+  }
+  if (operation.variables === undefined) {
+    operation.variables = {};
+  }
+  if (operation.query === undefined) {
```
Comment (jbaxleyiii) [inline src/link.ts]: what is this for?

File: src/link.ts [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+          types {
+            name
+          }
+        }
+      }
+    `;
+  }
+  const _operation = transformOperation(operation);
+
+  return link.request(_operation) || Observable.of();
```
Comment (jbaxleyiii) [inline src/link.ts]: should there be a `forward` passed to this request?

File: src/cacheLink.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
+import {
```
Comment (stubailo) [inline src/cacheLink.ts]: This file is unused - can we move it into the tests, or an unused folder, or remove it?

Comment (stubailo) [inline src/cacheLink.ts]: This file is unused - can we move it into the tests, or an unused folder, or remove it?

File: src/httpLink.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    });
+
+    this.requestLink = new SingleRequestLink({
+      ...(fetchParams || {}),
+      fetch: this._fetch,
+    });
+  }
+
+  public request(operation: Operation): Observable<FetchResult> | null {
+    this.headers = operation.context && operation.context.headers || {};
```
Comment (stubailo) [inline src/httpLink.ts]: For every link, we should make sure to document what it expects on the context.

File: src/httpUtils.ts [hunk 10]
```diff
@@ -0,0 +1,46 @@
+import {
```
Comment (stubailo) [inline src/httpUtils.ts]: This file is not used. Let's remove it?

File: src/link.ts [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+  validateOperation(operation);
+
+  if (!operation.context) {
+    operation.context = {};
+  }
+  if (!operation.variables) {
+    operation.variables = {};
+  }
+  if (!operation.query) {
+    console.warn(`query should either be a sting or GraphQL AST`);
```
Comment (stubailo) [inline src/link.ts]: `sting` -> `string`

File: src/link.ts [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+    operation.query = <DocumentNode>{};
+  }
+
+  const _operation: Operation = transformOperation(operation);
```
Comment (stubailo) [inline src/link.ts]: Maybe `transformedOperation`? `_operation` isn't a super descriptive name.

File: src/link.ts [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+  return link.request(_operation) || Observable.of();
+}
+
+const toPromise = (link: ApolloLink) => {
+  return (operation: Operation, forward?: NextLink) => {
+    const observable = link.request(operation, forward);
+
+    return new Promise((resolve, reject) => {
+      observable.subscribe({
+        next: resolve,
```
Comment (stubailo) [inline src/link.ts]: I think if the observable has multiple results this should print a warning

File: src/link.ts [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+    link = new FunctionLink(link);
+  }
+  return {
+    request: toPromise(link),
+  };
+}
+
+function transformOperation(operation) {
+  if (typeof operation.query === 'string') {
+    operation = {
```
Comment (stubailo) [inline src/link.ts]: It's odd that this is taking pains not to mutate the input, but the below code doesn't.

File: src/link.ts [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+      ...operation,
+      query: parse(operation.query),
+    };
+  }
+
+  if (!operation.operationName) {
+    operation.operationName = operation.query.definitions &&
```
Comment (stubailo) [inline src/link.ts]: This will fail on:

```graphql
fragment aFragmentDefinition on Query { lol }

query theQuery { ...lol }
```

File: src/link.ts [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+      operation.query.definitions.length &&
+      operation.query.definitions[0].name || '';
+  }
+
+  return operation;
+}
+
+export class TerminatedLink extends ApolloLink {
```
Comment (stubailo) [inline src/link.ts]: Maybe `TerminatingLink`

Comment (stubailo) [inline src/link.ts]: Also, I don't think this class is used right now, other than in a ternary expression where it doesn't provide any type safety.

File: src/link.ts [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+  constructor(private f: RequestHandler) {
+    super();
+  }
+
+  public request(operation: Operation): Observable<FetchResult> {
+    return this.f(operation) || Observable.of();
+  }
+}
+
+export class TerminatedConcat extends ApolloLink {
```
Comment (stubailo) [inline src/link.ts]: Same as above, this doesn't provide any extra type safety where it's used

File: src/link.ts [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+    if (ApolloLink.isTerminating(left) && ApolloLink.isTerminating(right)) {
+      return new TerminatedSplit(test, left, right);
+    }
+
+    return new SplitLink(test, left, right);
+  }
+
+  public static toLink(link: ApolloLink | RequestHandler): ApolloLink {
+    if (typeof link === 'function') {
+      return link.length <= 1 ? new TerminatedLink(link) : new FunctionLink(link);
```
Comment (stubailo) [inline src/link.ts]: If we want to keep these here, we should use `.isTerminating`. But I don't think this helps since the function returns an `ApolloLink` anyway.

File: src/link.ts [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+    right: ApolloLink | RequestHandler = ApolloLink.passthrough(),
+  ): ApolloLink {
+    return this.concat(<ApolloLink>ApolloLink.split(test, left, right));
+  }
+
+  // join two Links together
+  public concat(link: ApolloLink | RequestHandler): ApolloLink {
+    if (this.request.length <= 1) {
+      const warning = Object.assign(
+        new Error(`You are calling concat a terminating link, which will have no effect`),
```
Comment (stubailo) [inline src/link.ts]: `calling concat a` -> `calling concat on a`

File: src/link.ts [hunk 20]
```diff
@@ -0,0 +1,234 @@
...
+    left: ApolloLink | RequestHandler,
+    right: ApolloLink | RequestHandler = ApolloLink.passthrough(),
+  ): ApolloLink {
+    return this.concat(<ApolloLink>ApolloLink.split(test, left, right));
+  }
+
+  // join two Links together
+  public concat(link: ApolloLink | RequestHandler): ApolloLink {
+    if (this.request.length <= 1) {
+      const warning = Object.assign(
```
Comment (stubailo) [inline src/link.ts]: What is this assign for?

File: src/link.ts [hunk 21]
```diff
@@ -0,0 +1,234 @@
...
+        next: resolve,
+        error: reject,
+      });
+    });
+  };
+};
+
+export function asPromiseWrapper(link: ApolloLink | RequestHandler) {
+  if (typeof link === 'function') {
+    link = new FunctionLink(link);
```
Comment (stubailo) [inline src/link.ts]: Can't we use `toLink` here?

File: src/linkUtils.ts [hunk 22]
```diff
@@ -0,0 +1,41 @@
+import {
```
Comment (stubailo) [inline src/linkUtils.ts]: Should we move `toLink`, `isTerminating`, etc, here?

File: src/mockLink.ts [hunk 23]
```diff
@@ -0,0 +1,22 @@
+import {
```
Comment (stubailo) [inline src/mockLink.ts]: This is not used

File: src/pollingLink.ts [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+
+
+export default class PollingLink extends ApolloLink {
+
+  private pollInterval: number;
+  private timer;
+
+  constructor(pollInterval?: number) {
+    super();
+    this.pollInterval = pollInterval || 0;
```
Comment (stubailo) [inline src/pollingLink.ts]: I'd say we should just throw if the poll interval is not set. `0` doesn't seem like a reasonable default.

File: src/pollingLink.ts [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+      const subscriber = {
+          next: (data) => {
+            observer.next(data);
+          },
+          error: (error) => observer.error(error),
+      };
+
+      const poll = (() => {
+        forward(operation).subscribe(subscriber);
+      }).bind(this);
```
Comment (stubailo) [inline src/pollingLink.ts]: I don't think the bind here is necessary.

File: src/retryLink.ts [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+            }, this.interval(this.delay, this.count));
+          } else {
+            observer.error(error);
+          }
+        },
+        complete: () => {
+          observer.complete();
+        },
+        // This causes an error, not sure why
+        // complete: observer.complete
```
Comment (stubailo) [inline src/retryLink.ts]: You probably need `observer.complete.bind(observer)`.

File: src/pollingLink.ts (hunk 24 above)
Comment (evanshauser) [inline src/pollingLink.ts]: should be interval

File: src/setContextLink.ts [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+    if (!operation.context) {
+      operation.context = {};
+    }
+    operation.context = {
```
Comment (evanshauser) [inline src/setContextLink.ts]: This should be a recursive merge, since a link could overwrite an entire object, such as headers

File: src/linkUtils.ts [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+  const OPERATION_FIELDS = ['query', 'operationName', 'variables', 'context'];
+  for (let key of Object.keys(operation)) {
+    if (OPERATION_FIELDS.indexOf(key) < 0) {
+      throw new Error(`illegal argument: ${key}`);
+    }
+  }
+}
+
+export function ensureForward(forward?: NextLink): void {
+  if (!forward || typeof forward !== 'function') {
```
Comment (jbaxleyiii) [inline src/linkUtils.ts]: the type check here should be enough https://runkit.com/jbaxleyiii/596d174b03c14800141f7fcf

File: src/pollingLink.ts (hunk 24 above)
Comment (jbaxleyiii) [inline src/pollingLink.ts]: @evanshauser are you open to it being a function that given the operation, can determine the interval?

File: src/pollingLink.ts [hunk 29]
```diff
@@ -0,0 +1,49 @@
...
+
+      if (this.pollInterval) {
+        this.timer = setInterval(poll, this.pollInterval);
+      }
+
+      forward(operation).subscribe(subscriber);
+
+      return () => {
+        clearInterval(this.timer);
```
Comment (jbaxleyiii) [inline src/pollingLink.ts]: I think you would also need to clean up the subscribers that have been created during polling yes? Since each interval will create a new subscriber

Comment (jbaxleyiii) [inline src/pollingLink.ts]: also, we need to check to make sure `this.timer` exists. I would probably just use a local variable instead of attaching it to the instance as well.

File: src/retryLink.ts [hunk 30]
```diff
@@ -0,0 +1,66 @@
...
+    interval?: (delay: number, count: number) => number,
+  }) {
+    super();
+    this.max = params && params.max || 10;
+    this.delay =  params && params.delay || 300;
+    this.interval =  params && params.interval || this.defaultInterval;
+  }
+
+  public request(operation: Operation, forward: NextLink): Observable<FetchResult> {
+    ensureForward(forward);
```
Comment (jbaxleyiii) [inline src/retryLink.ts]: should this be attached to the Link interface so it can be used without importing? The forward need is still I think the confusingish part here, so lets try to make it an easy thing for library developers to use

File: src/retryLink.ts [hunk 31]
```diff
@@ -0,0 +1,66 @@
...
+
+
+    return new Observable(observer => {
+      const subscriber = {
+        next: data => {
+          observer.next(data);
+          this.count = 0;
```
Comment (jbaxleyiii) [inline src/retryLink.ts]: I would switch these in case observer.next throws, we wan't the count reset anytime `.next` is called

File: src/retryLink.ts [hunk 32]
```diff
@@ -0,0 +1,66 @@
...
+        },
+        error: error => {
+          this.count++;
+          if (this.count < this.max) {
+            setTimeout(() => {
```
Comment (jbaxleyiii) [inline src/retryLink.ts]: lets store the timer so it can be cleaned up if it needs to be

File: src/retryLink.ts [hunk 33]
```diff
@@ -0,0 +1,66 @@
...
+          }
+        },
+        complete: () => {
+          observer.complete();
+        },
+        // This causes an error, not sure why
+        // complete: observer.complete
+      };
+
+      forward(operation).subscribe(subscriber);
```
Comment (jbaxleyiii) [inline src/retryLink.ts]: We should return an unsubscriber here that cleans up subscriptions and timers

File: src/retryLink.ts [hunk 34]
```diff
@@ -0,0 +1,66 @@
...
+    });
+  }
+
+  private defaultInterval = (delay, count) => delay;
```
Comment (jbaxleyiii) [inline src/retryLink.ts]: we don't need to access `count` if not being used

File: src/setContextLink.ts (hunk 27 above)
Comment (jbaxleyiii) [inline src/setContextLink.ts]: fwiw, recursive merges are really expensive :( I would say that this link overwrites and previous context of the same keys when being added.

@evanshauser when we call request within the `ApolloLink` interface, thoughts on creating a copy of the operation and passing it down?

File: src/singleRequestLink.ts [hunk 35]
```diff
@@ -0,0 +1,51 @@
...
+  createApolloFetch,
+  ApolloFetch,
+} from 'apollo-fetch';
+import { print } from 'graphql/language/printer';
+
+import {
+  ApolloLink,
+} from './link';
+
+export default class SingleRequestLink extends ApolloLink {
```
Comment (jbaxleyiii) [inline src/singleRequestLink.ts]: how does this differ from `httpLink`? Why would I use this over that?

File: src/singleRequestLink.ts [hunk 36]
```diff
@@ -0,0 +1,51 @@
...
+    fetch?: ApolloFetch,
+  }) {
+    super();
+    this._fetch = fetchParams && fetchParams.fetch || createApolloFetch({ uri: fetchParams && fetchParams.uri });
+  }
+
+  public request(operation: Operation): Observable<FetchResult> {
+    const request = {
+      ...operation,
+      query: print(operation.query),
```
Comment (jbaxleyiii) [inline src/singleRequestLink.ts]: this isn't needed since execute does it?

File: src/singleRequestLink.ts [hunk 37]
```diff
@@ -0,0 +1,51 @@
...
+    };
+
+    return new Observable<FetchResult>(observer => {
+      this._fetch(request)
+      .then(data => {
```
Comment (jbaxleyiii) [inline src/singleRequestLink.ts]: *nit spacing here is off

File: src/types.ts [hunk 38]
```diff
@@ -0,0 +1,49 @@
+import { DocumentNode } from 'graphql/language/ast';
+import * as Observable from 'zen-observable';
+
+export interface GraphQLRequest {
+  //This string could be an id for a persisted query
```
Comment (jbaxleyiii) [inline src/types.ts]: lets remove this comment as I don' think it is correct

File: src/types.ts [hunk 39]
```diff
@@ -0,0 +1,49 @@
...
+  query?: string | DocumentNode;
+  variables?: object;
```
Comment (jbaxleyiii) [inline src/types.ts]: lets make this and context [a `Record` type](https://www.typescriptlang.org/play/index.html#src=let%20x%3A%20Record%3Cstring%2C%20any%3E%0D%0A%0D%0Ax%20%3D%20%7B%20a%3A%201%20%7D%0D%0Ax.b%20%3D%204%0D%0A%0D%0Alet%20y%3A%20Object%0D%0Ay%20%3D%20%7B%20a%3A%201%20%7D%0D%0Ay.b%20%3D%204)

File: src/types.ts [hunk 40]
```diff
@@ -0,0 +1,49 @@
...
+  context?: object;
+}
+
+export interface Link {
```
Comment (jbaxleyiii) [inline src/types.ts]: Is this used?

File: src/types.ts [hunk 41]
```diff
@@ -0,0 +1,49 @@
...
+  request (operation: Operation): Promise<FetchResult>;
+}
+
+export interface Operation {
+  query: DocumentNode;
+  variables?: {
+    [variables: string]: any;
+  };
+  operationName?: string;
+  context?: {
```
Comment (jbaxleyiii) [inline src/types.ts]: same comment for Record as above

File: src/types.ts [hunk 42]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+export interface FetchResult {
+  data: any;
+  errors?: any;
+  extensions?: any;
+  context?: object;
+}
+
+export interface Subscription {
```
Comment (jbaxleyiii) [inline src/types.ts]: Lets check to see if `@types/zen-observable` has these so we can use the types of the underlying observable implementation

File: tests/httpLink.ts [hunk 43]
```diff
@@ -0,0 +1,215 @@
...
+  });
+
+  afterEach(() => {
+    fetchMock.restore();
+  });
+
+  it('raises warning if called with concat', () => {
+    const link = ApolloLink.from([new HttpLink()]);
+    const _warn = console.warn;
+    console.warn = warning => assert.property(warning, 'message');
```
Comment (jbaxleyiii) [inline tests/httpLink.ts]: its been a while since I've used mocha, will this error if console.warn isn't called?

File: tests/httpLink.ts [hunk 44]
```diff
@@ -0,0 +1,215 @@
...
+    assert.deepEqual(link.concat((operation, forward) => forward(operation)), link);
+    console.warn = _warn;
+  });
+
+  it('does not need any constructor arguments', () => {
+    assert.doesNotThrow(() => new HttpLink());
```
Comment (jbaxleyiii) [inline tests/httpLink.ts]: shouldn't it? What would it do in that case?

File: tests/httpLink.ts [hunk 45]
```diff
@@ -0,0 +1,215 @@
...
+  });
+
+  it('calls next and then complete', (done) => {
+    const next = sinon.spy();
+    const link = new HttpLink({uri: 'data'});
+    const observable = link.request({
```
Comment (jbaxleyiii) [inline tests/httpLink.ts]: For all of these tests, I think we should be using `execute(link, operation)` instead of the `.request`. As link grows, I can see things being done during execution that each individual ink wouldn't need to ensure in its request (for instance mutation of the operation). 

I don't think `request` is really meant to be used publically on the link, instead it is used by the library itself to build the execution chain

File: tests/httpLink.ts [hunk 46]
```diff
@@ -0,0 +1,215 @@
...
+      query: sampleMutation,
+      context,
+      variables,
+    });
+    observable.subscribe({
+      next,
+      error: (error) => assert(false),
+      complete: () => {
+        const body = JSON.parse(fetchMock.lastCall()[1]['body']);
+        assert.equal(body['query'], print(sampleMutation));
```
Comment (jbaxleyiii) [inline tests/httpLink.ts]: any reason for not using dot notation here?

File: tests/link.ts [hunk 47]
```diff
@@ -0,0 +1,862 @@
...
+    id
+  }
+}`;
+
+describe('ApolloLink(abstract class)', () => {
+
+  describe('concat', () => {
+    it('should concat a function', (done) => {
+      const returnOne = new SetContextLink({ add: 1 });
+      const link = returnOne.concat((operation, forward) => Observable.of({ data: operation.context.add }));
```
Comment (jbaxleyiii) [inline tests/link.ts]: @evanshauser this shows my potential fear with forward :(

I'm assuming the library would not see this as a terminating link, yet forward is not used so it in essence is terminating. I think with copy paste / boilerplate, this could really get people confused

File: tests/link.ts [hunk 48]
```diff
@@ -0,0 +1,862 @@
...
+        operation = {
+          ...operation,
+          context: {
+            add: operation.context.add + 2,
+          },
+        };
+        return forward(operation);
+      });
+      const mock2 = new MockLink((op, forward) => Observable.of({ data: op.context.add + 2 }));
+      const mock3 = new MockLink((op, forward) => Observable.of({ data: op.context.add + 3 }));
```
Comment (jbaxleyiii) [inline tests/link.ts]: lets move `MockLink` to the `/tests` folder so it won't be confused as an external link

File: tests/singleRequestLink.ts [hunk 49]
```diff
@@ -0,0 +1,9 @@
+import { assert } from 'chai';
+import SingleRequestLink from '../src/singleRequestLink';
+
+
+describe('SingleRequestLink', () => {
```
Comment (jbaxleyiii) [inline tests/singleRequestLink.ts]: I don't think this is needed since the link won't be shipped. If it is, this needs a lot more tests

File: tests/tests.ts [hunk 50]
```diff
@@ -0,0 +1,7 @@
+import './httpLink';
```
Comment (jbaxleyiii) [inline tests/tests.ts]: Is this file needed? Can you not glob test?

File: src/link.ts [hunk 51]
```diff
@@ -0,0 +1,234 @@
...
+import { DocumentNode } from 'graphql/language/ast';
+
+export abstract class ApolloLink {
+
+  public static from(links: (ApolloLink | RequestHandler)[]) {
+    if (links.length === 0) {
+      return ApolloLink.empty();
+    }
+
+    return links
```
Comment (evanshauser) [inline src/link.ts]: @jbaxleyiii The issue I ran into was that reduce is applied, so that the initial argument is the first in the chain, so the function calls became:

```js
ApolloLink.empty().concat(first).concat(second)... 
```

We could do an ApolloLink.passthrough() as the initial value

File: src/pollingLink.ts (hunk 24 above)
Comment (evanshauser) [inline src/pollingLink.ts]: Sounds good!

File: src/link.ts (hunk 51 above)
Comment (jbaxleyiii) [inline src/link.ts]: ah shoot, yep! thank you! Lets keep it as is
